### ISSUE TITLE
KaTeX MathML

### ISSUE THREAD

**Issue by ronkok** [KaTeX/KaTeX#1876]
According to a recent [announcement](https://mathml.igalia.com/news/2019/02/12/launch-of-the-project/#new), there is work underway to enable MathML in Chromium. If all major browsers ever do support MathML, then a good TeX-to-MathML conversion library would be a valuable thing. I was curious to know if KaTeX could be the basis for such a library so I did some research.

You can see the results by using Firefox to view [this page](http://webmathcomparison.net/). It extracts KaTeX’s MathML output and passes it to the browser to be rendered. The results are better than I expected. And they suggest (at least to me) some fairly obvious next steps.

It would be good if KaTeX exposed MathML output. In my test page, I got the MathML by generating a KaTeX element and finding `el.children[0].children[0].children[0]`.

In the test page, it’s clear that some KaTeX MathML items can be improved: 
- [ ] Horizontal spacing between elements in environments
- [ ] `\cfrac` and `\dfrac` should be `\displaystyle`. Possibly also `\genfrac`.
- [ ] `{aligned}` alignment
- [ ] `\hline`, `\hdashline`, and vertical `{array}` lines
- [ ] `\brace`
- [ ] Spacing within the various `eqcolon` items.
- [ ] `\limits`
- [ ] `\overgroup`
- [ ] `\overset` and `\stackrel` and `\underset`
- [ ] `\pmb`
- [ ] `\underbrace`
- [ ] `\undergroup`
- [ ] `\underlinesegment`
- [ ] `\vphantom` needs a look.

**Comment by kevinbarabash** [comment]
Thanks for putting that comparison page together.  I tested a few of the items you called out in MathJax and it does a pretty good job.  I think there's a lot we can learn from its MathML output.

**Comment by ronkok** [comment]
I agree about the MathJax output, and I've changed the page so that MathJax writes native MathML, which makes it quite clear. And yet MathJax also has a couple of problems. Their `\oint` needs work. Something is also off on `\lim_y x`, though that might be a Firefox problem.

**Comment by flying-sheep** [comment]
Everything supports MathML now, and the link above is broken: http://webmathcomparison.net/

Would love to be able to set KaTeX to “just emit MathML, don’t even bother with CSS” mode.

**Comment by ronkok** [comment]
I also wanted direct MathML output, so I forked KaTeX and created the [Temml](https://temml.org/) library.

> don’t even bother with CSS

CSS is still necessary because the three main browser rendering engines are not consistent in how they support MathML. Temml contains several work-arounds in the service of getting good rendering.

I also wrote an updated [feature comparison](https://temml.org/docs/en/comparison).

**Comment by mbourne** [comment]
Hi Ron

Your interesting feature comparison page uses MathJax 2.7.5, but it's in version 4 now (that old version really clogs up the page rendering). There are some things indicated as "Not supported" or that don't appear to work but actually do on MathJax's own demo (like applying a class, \cancel,  \text{\~{a}}, array with lines, etc). Similarly with the KaTeX examples.

Regards
Murray

**Comment by ronkok** [comment]
That comparison page is meant to show what each library can emit as direct MathML output, not what they can emit in HTML. Older versions of MathJax had an option to emit only MathML with no parallel HTML output. I've tried and failed to get the same kind of purely MathML output from newer versions of MathJax. They did a major rewrite for version 3, which considerably sped up the rendering. But it is HTML output, not pure MathML.

I'll make another attempt to get pure MathML from the current version of MathJax, but I make no promises. Ideally, the script would work server-side instead of client-side. Even in MathJax 2.7.5, I was unable to get server-side pure MathML. That is why the page runs the quite slow MathJax 2.7.5 client-side script.

If you know how to get MathJax 4 to emit pure MathML, I would appreciate some coaching on how to achieve that. Bonus points for server-side rendering.

Similarly, KaTeX does a better job with its HTML output than it does with its MathML output. The comparison page shows only MathML which accounts for the differences you have mentioned. It's possible that my comparison page has errors and I'll do some more review, but don't expect much KaTeX improvement purely due to that review.

**Comment by ronkok** [comment]
Good news! I just got MathJax 4 to emit MathML for the first time. I'll investigate further and I may be able to remove some of those `not supported` notes in the comparison page.

**Comment by ronkok** [comment]
I've gone through the comparison page after updating MathJax. I was able to eliminate ten instances of MathJax `not supported` and I was also able to change several MathJax `not supported` notes to `requires upgreek package`.

I've also updated the KaTeX on that page, but it did provide support for any new functions.

Differences between the comparison page and your typical MathJax or KaTeX output is now the difference between their HTML output and their MathML output.

@mbourne Thank you for the nudge to do this exercise.

**Comment by mbourne** [comment]
Yes, I'm aware your page is comparing MathML output. I'm wondering if it might be an idea for the table `<th>`s on to put "MathJax MathML", "KaTeX MathML" so it's a bit clearer for those who are not sure what's happening (and believe that MathJax's and KaTeX's alignment for the `\begin{align}...` environment is pretty bad, for example).

Not sure if you're aware, but the old call to MathJax 2.7.5 is still present at the bottom of the page, and may (or may not) be causing some grief.

**Comment by ronkok** [comment]
Thank you for alerting me to the call to version 2.7.5. The effect is interesting. Removing that call fixes some MathJax problems and causes some new ones. Your suggestion for column headings is also a good one.

I'll clean it up and republish the page tomorrow.

**Comment by ronkok** [comment]
I've updated the comparison table. It should now be a pretty accurate view of the current state of affairs.

**Comment by mbourne** [comment]
Looks good, Ron. Some minor points:

(1) I see quite a few red renderings in the Mathjax column. I'm not sure what the difference is between "Not supported" and a red rendering.

(2) The "code" column for the row `\textdollar` has a typo.

(3) All 3 renderings for `\genfrac` have really thick vincula. I know they're prone to disappear, but each seems somewhat excessive.

**Comment by ronkok** [comment]
Thanks Murray,

1. "Not supported" is what I have hand-written to avoid a garish extra-wide MathJax error message. With MathJax 4, it is now also writing some less garish red error messages for non-supported function names. This table has been pretty tedious to prepare and I haven't the energy to make all the messages consistent.

2. I've fixed the `\textdollar`. Thanks for the tip.

3. With `\genfrac`, the writer can control the thickness of the vinculum. My choice of thickness is not meant to suggest an aesthetic preference, but to make it obvious that such control exists.
